proxy/dialer.go

```diff
@@ -0,0 +1,53 @@
+package core
+
+import (
+	"net"
+	"errors"
+	"github.com/shadowsocks/go-shadowsocks2/socks"
+	"github.com/shadowsocks/go-shadowsocks2/core"
+)
+
+type ProxyDialer struct {
+	core.StreamConnCipher
+	addr string
+}
+
+func NewDialer(addr string, algo string, password string) (s *ProxyDialer, err error) {
+	cipher, err := core.PickCipher(algo, nil, password)
+	if err != nil {
+		return
+	}
+	s = new(ProxyDialer)
+	s.StreamConnCipher = cipher
+	s.addr = addr
+	return
+}
+
+func NewDialerFromKey(addr string, algo string, key []byte) (s *ProxyDialer, err error) {
+	cipher, err := core.PickCipher(algo, key, "")
+	if err != nil {
+		return
+	}
+	s = new(ProxyDialer)
+	s.StreamConnCipher = cipher
+	s.addr = addr
+	return
+}
+
+func (dialer *ProxyDialer) Dial(network, addr string) (c net.Conn, err error) {
```

Comment:
You should call this DialTCP, as in [net.DialTCP](https://golang.org/pkg/net/#DialTCP).
Unless you also support UDP or other networks.

In any case, you should be checking the value of the `network` parameter.